### DISCUSSION TITLE
Disconnect peers that does not support a feature

### DISCUSSION THREAD

**Discussion by sondreb** [block-core/blockcore#336] [General]
On the `OnPartialValidationCompleted`, if the network has activated some BIP9 deployment, we will proceed to disconnect nodes that does not support the feature we need.

That means as soon as a blockchain network activates, it will proceed to disconnect those nodes that does not support, e.g. Segwit.

Recent improvement to this code ensures that the nodes will always announce Segwit-capability, but this does not apply to older versions of the software, so they gets disconnected.

Code:

```
        /// <inheritdoc />
        public void AddDiscoveredNodesRequirement(NetworkPeerServices services)
        {
            IPeerConnector peerConnector = this.PeerConnectors.FirstOrDefault(pc => pc is PeerConnectorDiscovery);
            if ((peerConnector != null) && !peerConnector.Requirements.RequiredServices.HasFlag(services))
            {
                peerConnector.Requirements.RequiredServices |= services;
                foreach (INetworkPeer peer in peerConnector.ConnectorPeers)
                {
                    if (peer.Inbound)
                        continue;

                    if (!peer.PeerVersion.Services.HasFlag(services))
                        peer.Disconnect("The peer does not support the required services requirement.");
                }
            }
        }
```

Should we change this logic to be similar to proven headers, where an % of connections that doesn't support PH, will still be kept around?

If anyone is deploying with flags, they should consider dropping the last logic of the code above in the file `ConnectionManager.cs`. Just disable the if check and peer.Disconnect.

**Comment by dangershony** [comment]
> Recent improvement to this code ensures that the nodes will always announce Segwit-capability, but this does not apply to older versions of the software, so they gets disconnected. 

Just because of this comment I think we should not make this change because I am not sure the implications of that and if new versions fix it then it will go away pretty soon.
